Fix replica-primary inconsistencies when indexing during primary relocation with ongoing replica recoveries

ywelsch opened the pull request:
Primary relocation violates two invariants that ensure proper interaction between document replication and peer recoveries, ultimately leading to documents not being properly replicated (see #19248 for more details).

Closes #19248

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java:
```diff
@@ -152,6 +160,51 @@ public boolean cancelRecoveriesForShard(ShardId shardId, String reason) {
+                }
+
+                @Override
+                public void onClusterServiceClose() {
```
we use this implementation a lot - maybe we should make utility base class for it (different PR). a default method implementation will be tricky because of the local node. Maybe it should be a parameter to the onClusterServiceClose. Food for thought.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java:
```diff
@@ -119,16 +114,7 @@ private RecoveryResponse recover(final StartRecoveryRequest request) throws IOEx
         }
 
         logger.trace("[{}][{}] starting recovery to {}", request.shardId().getIndex().getName(), request.shardId().id(), request.targetNode());
```
can we log this after the add(), which may actually delay the recovery?

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java:
```diff
@@ -144,38 +130,57 @@ public void messageReceived(final StartRecoveryRequest request, final TransportC
+        private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
+
+        synchronized RecoverySourceHandler add(StartRecoveryRequest request, IndexShard shard) {
+            final RecoverySourceHandler handler;
```
maybe just :

```
final ShardRecoveryContext shardContext = ongoingRecoveries.computeIfAbsent(shard, s -> new ShardRecoveryContext());
return shardContext.addNewRecovery(() -> createRecoverySourceHandler(request, shard, shardContext));
```

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java:
```diff
@@ -144,38 +130,57 @@ public void messageReceived(final StartRecoveryRequest request, final TransportC
-            if (shardRecoveryHandlers == null) {
-                shardRecoveryHandlers = new HashSet<>();
-                ongoingRecoveries.put(shard, shardRecoveryHandlers);
+        private RecoverySourceHandler createRecoverySourceHandler(StartRecoveryRequest request, IndexShard shard,
```
can we move this to ShardRecoveryContext and pass request and shard to addNewRecovery ? this way won't need all the supplier fun. Might as well rename the add method here to addNewRecovery too? or maybe `registerRecovery`?  (I don't mind much about the names, just a suggestion)

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java:
```diff
@@ -144,38 +130,57 @@ public void messageReceived(final StartRecoveryRequest request, final TransportC
             }
-            if (shardRecoveryHandlers.isEmpty()) {
+            if (shardRecoveryContext.recoveryHandlers.isEmpty()) {
                 ongoingRecoveries.remove(shard);
```
can we assert that the delay recovery exception is null?

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -136,6 +146,16 @@ public RecoveryResponse recoverToTarget() throws IOException {
+            // engine was just started at the end of phase 1
+            if (shard.state() == IndexShardState.RELOCATED) {
+                /**
+                 * we could have missed replication operations on the target if its engine wasn't started yet after primary relocation
```
maybe, "the primary shard has been relocated while we copied files. This means we can't guarantee any more that any operation that was replicated during the file copy (when the target engine was not yet opened) will be present in the local translog and thus will be resent on phase 2. The reason is that operation replicated by the target primary are sent to recovery target and the local shard (old primary) concurrently, meaning it may have arrived at the recovery target before we opened the engine and is still on inlight on the local shard.

Checking the relocated status here, after we opened the engine on the target,  is safe because primary relocation waits for all ongoing operations to complete and be fully replicated. Therefore all future operation by the new primary are guaranteed to reach the target shard when it's engine is open.
"

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -136,6 +146,16 @@ public RecoveryResponse recoverToTarget() throws IOException {
+                 * hand-off completed. The reason is that replication failures are just (silently) ignored if the engine isn't started yet,
+                 * see {@link ReplicationOperation#ignoreReplicaException(Throwable)} and {@link IndexShard#getEngine()}.
+                 */
+                throw new IndexShardRelocatedException(request.shardId());
```
do we need to use IndexShardRelocationException? I looked at the other usages of it, and I think we can get of it all together? We can  use IllegalIndexShardStateException ?

ywelsch replied:
We talked about this already on another PR, and my answer is the same here: We currently have an exception for each shard state. Some of the others are used as sparsely as this one. If we want to clean this up, we should not only remove this one. Can be a follow-up but it's really low on my list...

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -362,11 +382,37 @@ public void finalizeRecovery() {
+            // replica shards that have recovered or are still recovering from the current primary, otherwise replication actions
+            // will not be send to these replicas. To accomplish this, first block new recoveries, then take version of latest cluster
+            // state. This means that no new recovery can be completed based on information of a newer cluster state than the current one.
+            newRecoveriesHandler.handle(new DelayRecoveryException("primary relocation hand-off in progress or completed for " + shardId));
```
I think we should make newRecoveriesHanlder be:
```
 private final Function<String, Releasable> delayNewRecoveries; 
```

and then all of this becomes:

```
            try (Releasable ignored = delayNewRecoveries.apply("primary relocation hand-off in progress or completed for " + shardId)) {
                final long currentClusterStateVersion = currentClusterStateVersionSupplier.get();
                logger.trace("[{}][{}] waiting on {} to have cluster state with version [{}]", indexName, shardId, request.targetNode(),
                    currentClusterStateVersion);

                cancellableThreads.execute(() -> recoveryTarget.waitForClusterState(currentClusterStateVersion));

                logger.trace("[{}][{}] performing relocation hand-off to {}", indexName, shardId, request.targetNode());
                cancellableThreads.execute(() -> shard.relocated("to " + request.targetNode()));
            }
```

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:
```diff
@@ -85,7 +87,8 @@
     private final Map<String, String> tempFileNames = ConcurrentCollections.newConcurrentMap();
 
-    public RecoveryTarget(IndexShard indexShard, DiscoveryNode sourceNode, RecoveryTargetService.RecoveryListener listener) {
+    public RecoveryTarget(IndexShard indexShard, DiscoveryNode sourceNode, RecoveryTargetService.RecoveryListener listener,
```
can we add some java docs?

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetHandler.java:
```diff
@@ -44,6 +44,11 @@
     /**
+     * Blockingly waits for cluster state with at least clusterStateVersion to be available
+     */
+    void waitForClusterState(long clusterStateVersion);
```
nit: maybe `ensureClusterStateVersion` ?

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveryWaitForClusterStateRequest.java:
```diff
@@ -0,0 +1,70 @@
+
+public class RecoveryWaitForClusterStateRequest extends TransportRequest {
+
+    private long recoveryId;
```
time for a base class for all these recovery requests? we can remove a lot of boilerplate in RecoveryTargetService by have the  recovery resolving shared. As a different PR , of course.

bleskes commented:
thx @ywelsch . I left some comments. Can we also maybe add some tests to RecoverySourceHandlerTests to test the new relocation behavior (waiting for a cluster state version and fail on move to relocated )?

ywelsch commented:
@bleskes I've pushed a new change addressing comments. I've also added the unit tests that you asked for but I am not convinced that they are too useful.

bleskes commented:
@ywelsch I'm wondering if changing the title to `Fix replica-primary inconsistencies when indexing during primary relocation with ongoing replica recoveries` will better describe the situation?  or even take over the issue's title? (we use PR to drive change lists)

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java:
```diff
@@ -200,32 +177,50 @@ synchronized void cancel(IndexShard shard, String reason) {
-            synchronized void throwExceptionOnNewRecovery(@Nullable DelayRecoveryException onNewRecoveryException)
-                throws IllegalStateException {
-                if (onNewRecoveryException != null && this.onNewRecoveryException != null) {
+            synchronized Releasable blockNewRecoveries(String exceptionMessage) throws IllegalStateException {
```
nit: block suggest "block the thread" like on have it wait on a lock. don't you like "delay"?

ywelsch replied:
`delayNewRecoveries` sounds good!

bleskes commented:
LGTM. Thanks @ywelsch
